wagtail/documents/views/documents.py

```diff
@@ -157,117 +156,120 @@ def get_context_data(self, **kwargs):
-        )
-        if form.is_valid():
-            doc = form.save()
-
-            edit_url = reverse("wagtaildocs:edit", args=(doc.id,))
-            redirect_url = "wagtaildocs:index"
-            if next_url:
-                edit_url = f"{edit_url}?{urlencode({'next': next_url})}"
-                redirect_url = next_url
-
-            messages.success(
-                request,
-                _("Document '%(document_title)s' updated")
-                % {"document_title": doc.title},
-                buttons=[messages.button(edit_url, _("Edit"))],
-            )
-            return redirect(redirect_url)
-        else:
-            messages.error(request, _("The document could not be saved due to errors."))
-    else:
-        form = DocumentForm(instance=doc, user=request.user)
-
-    try:
-        local_path = doc.file.path
-    except NotImplementedError:
-        # Document is hosted externally (eg, S3)
-        local_path = None
-
-    if local_path:
-        # Give error if document file doesn't exist
-        if not os.path.isfile(local_path):
-            messages.error(
-                request,
-                _(
-                    "The file could not be found. Please change the source or delete the document"
-                ),
-                buttons=[
-                    messages.button(
-                        reverse("wagtaildocs:delete", args=(doc.id,)), _("Delete")
-                    )
-                ],
-            )
+    @cached_property
+    def model(self):
+        # Use a property instead of setting this as a class attribute so it is
+        # accessed at request-time, thus can be tested with override_settings
+        return get_document_model()
 
-    return TemplateResponse(
-        request,
-        "wagtaildocs/documents/edit.html",
-        {
-            "document": doc,
-            "filesize": doc.get_file_size(),
-            "form": form,
-            "user_can_delete": permission_policy.user_has_permission_for_instance(
-                request.user, "delete", doc
-            ),
-            "next": next_url,
-        },
-    )
+    def get_form_class(self):
+        return get_document_form(self.model)
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs["user"] = self.request.user
+        return kwargs
+
+    def get_initial_form_instance(self):
+        return self.model(uploaded_by_user=self.request.user)
+
+    def get_success_message(self, instance):
+        return _("Document '%(document_title)s' added.") % {
+            "document_title": instance.title
+        }
+
+
+class EditView(generic.EditView):
+    permission_policy = permission_policy
+    pk_url_kwarg = "document_id"
+    error_message = gettext_lazy("The document could not be saved due to errors.")
+    template_name = "wagtaildocs/documents/edit.html"
+    index_url_name = "wagtaildocs:index"
+    edit_url_name = "wagtaildocs:edit"
+    delete_url_name = "wagtaildocs:delete"
```

Comment:
Would be nice to wrap the views in a viewset, and maybe get the viewset class from the app config similar to wagtailusers so it can be customised, but I think that's out of scope for now.